src/ocaml_lsp.ml

```diff
@@ -53,64 +53,76 @@ type t =
+  let prefix = (Ocaml_version.major version, Ocaml_version.minor version) in
+  List.Assoc.find lsp_versions ~equal:Poly.equal prefix
 
 let is_version_up_to_date t ocaml_v =
-  let ocamllsp_version = get_version_semver t in
+  let ocamllsp_version = get_version_from_serverInfo t in
   let res =
+    (* if the server doesn't have a version, we assume it's ancient and we can
+       offer an upgrade *)
     match ocamllsp_version with
-    | Ok v -> (
-      match ocaml_v with
-      | _ when Ocaml_version.(ocaml_v < Releases.v4_06_0) ->
-        Error (`Ocaml_version_not_supported ocaml_v)
-      | _ when Ocaml_version.(ocaml_v < Releases.v4_12_0) ->
-        Ok (String.equal v "1.4.1")
-      | _ when Ocaml_version.(ocaml_v < Releases.v4_13_0) ->
-        Ok (String.equal v "1.9.0")
-      | _ when Ocaml_version.(ocaml_v < Releases.v4_14_0) ->
-        Ok (String.equal v "1.9.1")
-      | _ -> Error (`Ocaml_version_not_supported ocaml_v))
-    | Error e -> Error (`Unexpected e)
+    | None -> (
+      match available_versions ocaml_v with
+      | Some v -> Error (`Newer_available (None, Array.last v))
+      | None -> Ok ())
+    | Some v -> (
+      match available_versions ocaml_v with
+      | None -> Ok ()
+      | Some available -> (
+        match Array.findi available ~f:(fun _ -> String.equal v) with
+        | None -> Ok ()
+        | Some (pos, _) ->
+          let last = Array.length available - 1 in
+          if Int.equal pos last then Ok ()
+          else Error (`Newer_available (Some v, available.(pos)))))
   in
   Result.map_error res ~f:(fun err ->
       let msg =
         match err with
-        | `Ocaml_version_not_supported v ->
+        | `Newer_available (old, new_) ->
+          let upgrade =
+            match old with
+            | None -> sprintf "to %s" new_
+            | Some old -> sprintf "%s -> %s" old new_
+          in
           sprintf
-            "The installed version of OCaml is not supported by `ocamllsp`: \
-             %s. Consider upgrading OCaml version used in the current sandbox."
-            (Ocaml_version.to_string v)
-        | `Unexpected `Language_server_isn't_ocamllsp ->
-          "Using a language server besides `ocamllsp` isn't expected by this \
-           extension. Please, switch to using `ocamllsp` by installing the \
-           package `ocaml-lsp-server` in your current sandbox."
-        | `Unexpected `Missing_serverInfo
-        | `Unexpected `ServerInfo_version_missing ->
-          "The extension expected the server version to be sent by `ocamllsp`. \
-           It is missing. Please, consider upgrading the package \
-           `ocaml-lsp-server`."
-        | `Unexpected `Unable_to_parse_version ->
-          "The extension was unable to parse `ocamllsp` version. That's \
-           strange. Consider filing an issue on the project GitHub with the \
-           version of your `ocamllsp`."
+            "The current version of ocamllsp is out of date. Consider \
```

Comment:
```suggestion
            "There is a newer version of `ocamllsp` available. Consider \
```

"out of date" means it may not be supported already?

Reply:
Can you confirm that we can use markdown in the messages? I believe you said it was plaintext only earlier.